kernel: implement k_thread_create() as a syscall

tyrantlizard77 opened the pull request:
- for purposes of permission tracking, thread stacks are now treated like other kernel objects
- userspace handlers can now assert that objects passed in are strictly uninitialized
- added k_thread_create() as a system call

Commit:
kernel: implement k_thread_create() as a syscall

User threads can only create other nonessential user threads
of equal or lower priority and must have access to the entire
stack area.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented on kernel/thread.c:
```diff
@@ -290,7 +290,45 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+	_SYSCALL_MEMORY(stack, stack_size + guard_size, 1, ssf);
+
+	/* User threads may only create other user threads */
+	_SYSCALL_VERIFY(margs->arg9 & K_USER, ssf);
```
There's a race condition here: if another thread is furiously flipping this bit, there's a possibility that a thread might be spawn with higher privileges from another user thread. I don't see thread creation as being something performance- or latency- sensitive, so I think it's a better idea to do this in a critical section where no other thread is allowed to run during thread creation.

andrewboie replied:
ah good catch. yeah for arguments passed in via struct, and not registers, we can have attacks like this.
I'll lock interrupts.

andrewboie commented:
CI is failing due to some dependencies in my other system call PR. Rather than shuffle code around we'll just deal with that once the other PR goes in.

agross-oss reviewed (commented):
nit: a commit message describing the change would be good.

andrewboie commented:
Rebased, added to commit message

andrewboie commented on kernel/thread.c:
```diff
@@ -324,7 +324,54 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+
+	/* User threads may only create other user threads */
+	_SYSCALL_VERIFY(margs->arg9 & K_USER, ssf);
+
```
We need to verify that K_ESSENTIAL isn't passed in, user threads can't create essential threads

lpereira replied:
Good catch.  Need a whitelist of flags.

andrewboie commented on kernel/thread.c:
```diff
@@ -324,7 +324,54 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+	_SYSCALL_VERIFY(_is_prio_lower_or_equal(margs->arg8,
+						_current->base.prio), ssf);
+
+	ret = (u32_t)_impl_k_thread_create((struct k_thread *)new_thread,
```
need a check somewhere that we haven't run out of thread_ids

andrewboie commented:
new patch: rebased and updated to reflect current state of policy discussions

lpereira commented on kernel/thread.c:
```diff
@@ -305,7 +305,60 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+	_SYSCALL_VERIFY(_is_prio_lower_or_equal(prio, _current->base.prio));
+
+	/* Make sure we haven't run out of thread IDs */
+	_SYSCALL_VERIFY(thread_index < 8 * CONFIG_MAX_THREAD_BYTES);
```
Even though thread_index is an atomic variable, `atomic_inc()` will only guarantee the atomicity of the increment operation. This won't prevent a race condition from happening when this value is about to hit its limits (so it passes the check), and another thread is successfully created.

andrewboie commented:
updated: put an irq_lock() around the thread_index check

lpereira commented on kernel/thread.c:
```diff
@@ -305,7 +305,63 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+	_SYSCALL_VERIFY(_is_prio_lower_or_equal(prio, _current->base.prio));
+
+	/* Make sure we haven't run out of thread IDs */
+	key = irq_lock();
```
The IRQ lock here won't help.  Things can happen between irq_unlock() and whatever happens inside _impl_k_thread_create().

andrewboie replied:
ohhh I get it now.
you're right, bleah.

Commit:
userspace: treat thread stacks as kernel objects

We need to track permission on stack memory regions like we do
with other kernel objects. We want stacks to live in a memory
area that is outside the scope of memory domain permission
management. We need to be able track what stacks are in use,
and what stacks may be used by user threads trying to call
k_thread_create().

Some special handling is needed because thread stacks appear as
variously-sized arrays of struct _k_thread_stack_element which is
just a char. We need the entire array to be considered an object,
but also properly handle arrays of stacks.

Validation of stacks also requires that the bounds of the stack
are not exceeded. Various approaches were considered. Storing
the size in some header region of the stack itself would not allow
the stack to live in 'noinit'. Having a stack object be a data
structure that points to the stack buffer would confound our
current APIs for declaring stacks as arrays or struct members.
In the end, the struct _k_object was extended to store this size.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace handlers: finer control of init state

We also need macros to assert that an object must be in an
uninitialized state. This will be used for validating thread
and stack objects to k_thread_create(), which must not be already
in use.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
significantly re-worked; thread stacks are now treated as kernel objects.

lpereira commented on kernel/include/syscall_handler.h:
```diff
@@ -17,6 +17,12 @@
 
 extern const _k_syscall_handler_t _k_syscall_table[K_SYSCALL_LIMIT];
 
+enum _obj_init_check {
```
Since this is used inside the kernel object validation function, it might be a good idea to define these values to be -1, 0, and 1.  Integer comparison is already cheaper, but not only 0 is special, checking for the sign bit is cheaper. I'd still make `_OBJ_INIT_TRUE = 0`, as that's the one that's more likely to be true.

andrewboie replied:
done

lpereira commented on kernel/userspace.c:
```diff
@@ -222,7 +226,8 @@ void k_object_access_all_grant(void *object)
+int _k_object_validate(struct _k_object *ko, enum k_objects otype,
+		       enum _obj_init_check init)
 {
 	if (!ko || (otype != K_OBJ_ANY && ko->type != otype)) {
```
Unrelated to the patch, but we might want to add `unlikely()` here for the error cases.

lpereira commented on kernel/thread.c:
```diff
@@ -352,7 +354,82 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
+			  options);
+
+	if (new_thread->base.perm_index == -1) {
+		k_thread_abort(new_thread);
```
Nice solution

lpereira reviewed (approved):
LGTM.  A few low-priority comments, though.

andrewboie commented:
updated with feedback from @lpereira 
Also fixed CI issue by upping the space reserved for the gperf code from 118 to 128, seems adding the extra field to struct _k_object pushed it over the edge for a few tests with a lot of kernel objects.

lpereira commented on kernel/userspace.c:
```diff
@@ -222,24 +226,39 @@ void k_object_access_all_grant(void *object)
-	if (!init && !(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
-		return -EINVAL;
+	switch (init) {
+	case _OBJ_INIT_TRUE:
```
You'll only take advantage of the new values if you rewrite this using `if` statements:

```C
if (likely(init == _OBJ_INIT_TRUE)) {
		/* Object MUST be intialized */
		if (unlikely(!(ko->flags & K_OBJ_FLAG_INITIALIZED))) {
			return -EINVAL;
		}
} else if (init < _OBJ_INIT_TRUE) {
		/* Object MUST NOT be initialized */
		if (unlikely(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
			return -EADDRINUSE;
		}
} else {
                /* Don't care. */
}

return 0;
```

Don't take my word for it, see the [compiler output](https://godbolt.org/g/YzUUdp).

andrewboie replied:
fixed
